Engine/Common/Types.h

```diff
@@ -126,290 +126,557 @@ struct StartEndIndex {
+
+    // ---- comparison operators ---- //
+    // element by element comparison with operators //
+    DLLEXPORT constexpr bool operator<(const Int3& other) const noexcept
+    {
+        return std::tie(X, Y, Z) < std::tie(other.X, other.Y, other.Z);
+    }
+
+    DLLEXPORT constexpr bool operator<=(const Int3& other) const noexcept
     {
-        return X + Y + Z + W;
+        return std::tie(X, Y, Z) <= std::tie(other.X, other.Y, other.Z);
+    }
+
+    DLLEXPORT constexpr bool operator>(const Int3& other) const noexcept
+    {
+        return std::tie(X, Y, Z) > std::tie(other.X, other.Y, other.Z);
+    }
+
+    DLLEXPORT constexpr bool operator>=(const Int3& other) const noexcept
+    {
+        return std::tie(X, Y, Z) >= std::tie(other.X, other.Y, other.Z);
+    }
+
+    DLLEXPORT constexpr bool operator==(const Int3& other) const noexcept
+    {
+        return X == other.X && Y == other.Y && Z == other.Z;
+    }
+
+    DLLEXPORT constexpr bool operator!=(const Int3& other) const noexcept
+    {
+        return X != other.X || Y != other.Y || Z != other.Z;
+    }
+
+    // ------------------ Functions ------------------ //
+    // getters //
+    DLLEXPORT constexpr int GetX() const noexcept
+    {
+        return X;
+    }
+
+    DLLEXPORT constexpr int GetY() const noexcept
+    {
+        return Y;
+    }
+
+    DLLEXPORT constexpr int GetZ() const noexcept
+    {
+        return Z;
+    }
+
+    // setters //
+    DLLEXPORT inline void SetX(int val)
+    {
+        X = val;
+    }
+
+    DLLEXPORT inline void SetY(int val)
+    {
+        Y = val;
+    }
+
+    DLLEXPORT inline void SetZ(int val)
+    {
+        Z = val;
+    }
+
+    // add all elements together //
+    DLLEXPORT constexpr int HAdd() const noexcept
+    {
+        return X + Y + Z;
+    }
+
+    // Add all elements together absoluted (abs()) //
+    DLLEXPORT inline unsigned int HAddAbs() const noexcept
+    {
+        return std::abs(X) + std::abs(Y) + std::abs(Z);
+    }
+
+    // getting min and max of objects //
+    DLLEXPORT constexpr Int3 MinElements(const Int3& other) const noexcept
+    {
+        return Int3(
+            X < other.X ? X : other.X, Y < other.Y ? Y : other.Y, Z < other.Z ? Z : other.Z);
+    }
+
+    DLLEXPORT constexpr Int3 MaxElements(const Int3& other) const noexcept
+    {
+        return Int3(
+            X > other.X ? X : other.X, Y > other.Y ? Y : other.Y, Z > other.Z ? Z : other.Z);
+    }
+
+    // value clamping //
+    DLLEXPORT constexpr Int3 Clamp(const Int3& min, const Int3& max) const noexcept
+    {
+        const Int3 minval = this->MinElements(max);
+        return min.MaxElements(minval);
+    }
+
+    // ----------------- Vector math ------------------- //
+    // dot product of the vectors //
+    DLLEXPORT constexpr int Dot(const Int3& val) const noexcept
+    {
+        return X * val.X + Y * val.Y + Z * val.Z;
+    }
+
+    DLLEXPORT constexpr Int3 Cross(const Int3& val) const
```

Comment:
I don't think these explicitly vector operations should be defined for the integer classes. I don't think that makes much sense. Is there a use for them?

Reply:
i left them for consistency mostly, since they make matematical sense for integers as well (tho i cannot really think of an use case for them off the top of my head)